src/binder/bind_node_visitor.cpp

```diff
@@ -78,141 +88,166 @@ void BindNodeVisitor::Visit(parser::SelectStatement *node, parser::ParseResult *
   node->SetSelectColumns(new_select_list);
   node->SetDepth(context_->GetDepth());
 
-  auto curr_context = context_;
   context_ = context_->GetUpperContext();
-  delete curr_context;
 }
 
 // Some sub query nodes inside SelectStatement
-void BindNodeVisitor::Visit(parser::JoinDefinition *node, parser::ParseResult *parse_result) {
+void BindNodeVisitor::Visit(common::ManagedPointer<parser::JoinDefinition> node,
+                            common::ManagedPointer<BinderSherpa> sherpa) {
   BINDER_LOG_TRACE("Visiting JoinDefinition ...");
   // The columns in join condition can only bind to the join tables
-  node->GetLeftTable()->Accept(this, parse_result);
-  node->GetRightTable()->Accept(this, parse_result);
-  node->GetJoinCondition()->Accept(this, parse_result);
+  node->GetLeftTable()->Accept(common::ManagedPointer(this).CastManagedPointerTo<SqlNodeVisitor>(), sherpa);
+  node->GetRightTable()->Accept(common::ManagedPointer(this).CastManagedPointerTo<SqlNodeVisitor>(), sherpa);
+  node->GetJoinCondition()->Accept(common::ManagedPointer(this).CastManagedPointerTo<SqlNodeVisitor>(), sherpa);
 }
 
-void BindNodeVisitor::Visit(parser::TableRef *node, parser::ParseResult *parse_result) {
+void BindNodeVisitor::Visit(common::ManagedPointer<parser::TableRef> node,
+                            common::ManagedPointer<BinderSherpa> sherpa) {
   BINDER_LOG_TRACE("Visiting TableRef ...");
-  node->TryBindDatabaseName(default_database_name_);
+  ValidateDatabaseName(node, sherpa);
+
   if (node->GetSelect() != nullptr) {
     if (node->GetAlias().empty()) throw BINDER_EXCEPTION("Alias not found for query derived table");
 
     // Save the previous context
     auto pre_context = context_;
-    node->GetSelect()->Accept(this, parse_result);
-    // Restore the previous level context
+    node->GetSelect()->Accept(common::ManagedPointer(this).CastManagedPointerTo<SqlNodeVisitor>(), sherpa);
+    // TODO(WAN): who exactly should save and restore contexts? Restore the previous level context
```

Comment:
Select statment can save and restore the context upon entering and exiting its visit

Reply:
Right now, every statement has no concept of BinderContext at all. Are you proposing that BinderContext is a subset of BinderSherpa?

Reply:
Each statement starts with a fresh context, and during the binding, it will fill in the tables and columns and their alias that is visible to this statement. Maybe the name QueryContext is more appropriate as the context does not persist throughout the lifetime of the binder but only the life span of one query/subquery. Where this context resides does not matter much, so it definitely can be a part of the binder sherpa.

Reply:
Does it make sense for it to be a part of the binder sherpa, if it is logically distinct and has a different lifetime?